### PR TITLE
feat: [CDS-72657]: Adding API to fetch a list of manifest that can be applied as primary

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Adding API to fetch a list of manifest that can be applied to primaryManifestRef Field.
Screenshot 2023-06-27 at 7 41 28 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49715)

### CODE REVIEW

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java [hunk 1]
```diff
@@ -302,4 +305,38 @@ public static void populateIdentifiersOfEachLevel(String accountIdentifier, Stri
...
+    }
+    return manifestNode.getField(YamlTypes.MANIFEST_CONFIG);
+  }
+
+  public List<String> getManifestIdentifiersFilteredOnServiceType(YamlField manifestsField, String serviceType) {
+    List<String> manifestIdentifierList = new ArrayList<>();
+    List<YamlNode> manifests = manifestsField.getNode().asArray();
+    Set<String> supportedManifestTypes = ManifestType.getSupportedManifestTypes(serviceType);
+    for (YamlNode manifestConfig : manifests) {
+      YamlNode manifest = getManifestNodeYaml(manifestConfig).getNode();
```
Comment (TarunUba1) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:334]: similarly here just check ahead for null and then only fecth getNode, otherwise we will run in null.getNode() which feels weird

**Review by TarunUba1 — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java [hunk 2]
```diff
@@ -922,6 +923,35 @@ public ResponseDTO<ServiceInputsMergedResponseDto> mergeServiceInputs(
         serviceEntityService.validateTemplateInputs(accountId, orgId, projectId, serviceIdentifier, loadFromCache));
   }
 
+  @GET
```
Comment (abhinavsingh1196) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java:926]: this needs to be made using spec first approach and not code-first, since it'll be customer facing.
For this PR, you can just merge the underlying logic which gets manifest IDs and create the API layer later.

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java (hunk 1 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:334]: Done

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java [hunk 3]
```diff
@@ -302,4 +305,38 @@ public static void populateIdentifiersOfEachLevel(String accountIdentifier, Stri
       }
     }
   }
+
+  public YamlField getManifestsNodeFromServiceDefinitionYaml(YamlField serviceDefinitionField) {
+    if (serviceDefinitionField == null) {
```
[the remark this answers is not in the document]
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:310]: Not removing this as for all the null checks we are returning empty ManifestsResponseDTO object.

**Comment by GoelPratyush99** [comment]
trigger uts

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 4]
```diff
@@ -1415,6 +1415,57 @@ paths:
...
+    parameters:
+      - $ref: '#/components/parameters/ServicePathParam'
+      - $ref: '#/components/parameters/OrgPathParam'
+    get:
+      summary: List primary manifests
+      responses:
+        '200':
+          $ref: '#/components/responses/ManifestsResponseDTO'
+      operationId: get-org-scoped-primary-manifests
+      description: return a list of eligible primary manifests from all configured manifest in a org-scoped service.
```
Comment (TarunUba1) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:1446]: "in an org-scoped service" vowels

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 5]
```diff
@@ -1415,6 +1415,57 @@ paths:
...
+  '/v1/services/{service}/primaryManifests':
+    parameters:
+      - $ref: '#/components/parameters/ServicePathParam'
+    get:
+      summary: List primary manifests
+      responses:
+        '200':
+          $ref: '#/components/responses/ManifestsResponseDTO'
+      operationId: get-account-scoped-primary-manifests
+      description: return a list of eligible primary manifests from all configured manifest in a account-scoped service.
```
Comment (TarunUba1) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:1462]: "in an account-scoped"

**Comment by GoelPratyush99** [comment]
trigger uts

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 4 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: done

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 5 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: done

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 6]
```diff
@@ -1415,6 +1415,57 @@ paths:
       security:
         - x-api-key: []
       description: List rancher clusters using the given project level environment and infrastructure definition.
+  '/v1/orgs/{org}/projects/{project}/services/{service}/primaryManifests':
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:1418]: primary-manifests

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 7]
```diff
@@ -1415,6 +1415,57 @@ paths:
...
+      - $ref: '#/components/parameters/ServicePathParam'
+      - $ref: '#/components/parameters/OrgPathParam'
+      - $ref: '#/components/parameters/ProjectPathParam'
+    get:
+      summary: List primary manifests
+      responses:
+        '200':
+          $ref: '#/components/responses/ManifestsResponseDTO'
+      operationId: get-primary-manifests
+      description: return a list of eligible primary manifests from all configured manifest in a project-scoped service.
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:1429]: Returns ...

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 8]
```diff
@@ -2395,6 +2446,22 @@ components:
         allowed_for_api:
           type: boolean
           description: This indicates if a given IP is allowlisted in Harness for API requests
+    ManifestsResponseDTO:
+      title: ManifestsResponseDTO
+      type: object
+      properties:
+        manifestIdentifierList:
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:2453]: manifest_identifier_list

**Review by mankrit-singh — COMMENTED** [review]
Just some linting comments

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 9]
```diff
@@ -1415,6 +1415,57 @@ paths:
       security:
         - x-api-key: []
       description: List rancher clusters using the given project level environment and infrastructure definition.
+  '/v1/orgs/{org}/projects/{project}/services/{service}/primaryManifests':
+    parameters:
+      - $ref: '#/components/parameters/ServicePathParam'
+      - $ref: '#/components/parameters/OrgPathParam'
+      - $ref: '#/components/parameters/ProjectPathParam'
+    get:
+      summary: List primary manifests
```
Comment (abhinavsingh1196) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:1424]: can be `List primary manifests of a [project/org/account] scoped service`, and the same for other APIs as well, since this summary is displayed on the left nav IIRC

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 8 above)
Comment (abhinavsingh1196) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml:2453]: do we even need to say `list`? it could be `manifest_identifiers` or maybe even `identifiers`

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 10]
```diff
@@ -2395,6 +2446,22 @@ components:
...
+    ManifestsResponseDTO:
+      title: ManifestsResponseDTO
+      type: object
+      properties:
+        manifestIdentifierList:
+          type: array
+          description: List of Manifest Identifiers
+          items:
+            type: string
+      description: ManifestsResponseDTO return with a list of eligibe manifest identifiers that can be applied as primary manifest.
```
Comment (abhinavsingh1196) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:2458]: - typo in `eligible`
- would this be a better description?
`ManifestsResponseDTO returns a list of manifest identifiers from a service's eligible primary manifests`

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 11]
```diff
@@ -2395,6 +2446,22 @@ components:
...
+      x-internal: false
+      x-examples:
+        Example 1:
+          manifestIdentifierList:
```
Comment (abhinavsingh1196) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:2462]: will need modification accordingly

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java [hunk 12]
```diff
@@ -205,4 +205,21 @@ private List<ServiceResponse> filterByPermissionAndId(
     }
     return filteredAccessControlDtoList;
   }
+
+  public Response getPrimaryManifestList(String service, String org, String project, String harnessAccount) {
+    Optional<ServiceEntity> serviceEntity = serviceEntityService.get(harnessAccount, org, project, service, false);
```
Comment (abhinavsingh1196) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java:210]: nit: should have suffix `Optional`

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java [hunk 13]
```diff
@@ -205,4 +205,21 @@ private List<ServiceResponse> filterByPermissionAndId(
...
+
+    if (serviceEntity.isPresent()) {
+      if (EmptyPredicate.isEmpty(serviceEntity.get().getYaml())) {
+        throw new InvalidRequestException(
+            format("Service %s is not configured with a Service definition. Service Yaml is empty", service));
+      }
+      return Response.ok()
+          .entity(serviceEntityService.getManifestIdentifiers(serviceEntity.get().getYaml(), service))
+          .build();
+    } else {
```
Comment (abhinavsingh1196) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java:220]: throw this exception at the start i.e.

if (!serviceEntity.isPresent) { throw ...}
ServiceEntity service = serviceEntityOptional.get();
if (isEmpty(service.getYaml())) {throw ...}
return Response...

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java [hunk 14]
```diff
@@ -302,4 +305,40 @@ public static void populateIdentifiersOfEachLevel(String accountIdentifier, Stri
...
+    return manifestNode.getField(YamlTypes.MANIFEST_CONFIG);
+  }
+
+  public List<String> getManifestIdentifiersFilteredOnServiceType(YamlField manifestsField, String serviceType) {
+    List<String> manifestIdentifierList = new ArrayList<>();
+    List<YamlNode> manifests = manifestsField.getNode().asArray();
+    Set<String> supportedManifestTypes = ManifestType.getSupportedManifestTypes(serviceType);
+    for (YamlNode manifestConfig : manifests) {
+      YamlNode manifest = getManifestNodeYaml(manifestConfig).getNode();
+      if (manifest != null) {
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:335]: can just be `if (manifest != null && supportedManifestTypes.contains(manifest.getType()))`

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java [hunk 15]
```diff
@@ -302,4 +305,40 @@ public static void populateIdentifiersOfEachLevel(String accountIdentifier, Stri
...
+
+    YamlField serviceSpecField = serviceDefinitionField.getNode().getField(YamlTypes.SERVICE_SPEC);
+    if (serviceSpecField == null) {
+      return null;
+    }
+
+    return serviceSpecField.getNode().getField(YamlTypes.MANIFEST_LIST_CONFIG);
+  }
+
+  public YamlField getManifestNodeYaml(YamlNode manifestNode) {
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:322]: this can be private? and would be best to move it under the following method

File: 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java [hunk 16]
```diff
@@ -1164,6 +1167,34 @@ public void testMergeServiceInputsWithSingleArtifactSource() {
             + "                tag: <+input>\n");
   }
 
+  @Test
+  @Owner(developers = PRATYUSH)
+  @Category(UnitTests.class)
+  public void testGetManifestIdentifiersListServiceV2() {
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java:1173]: This test should be split into 3 tests.

Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java:1173]: Would it be better to mock ServiceFilterHelper for unit tests of ServiceEntityServiceImpl and have separate unit tests for ServiceFilterHelper? 
In any case, we should add tests for more negative scenarios for k8s/helm.

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 8 above)
Comment (mankrit-singh) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml:2453]: manifest_identifiers sounds good

**Comment by GoelPratyush99** [comment]
trigger uts

File: 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java (hunk 16 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java]: Done

Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java]: Done

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java (hunk 15 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java]: Done

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java (hunk 14 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java]: Done

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java (hunk 13 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java]: Done

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java (hunk 12 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java]: Done

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 11 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: Done

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 10 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: changed it

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 9 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: Done

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 8 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: changed it to identifiers

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 7 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: Done

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml (hunk 6 above)
Comment (GoelPratyush99) [reply 120-ng-manager/contracts/openapi/v1/openapi.yaml]: Done

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by abhinavsingh1196 — APPROVED** [review]

**Comment by GoelPratyush99** [comment]
trigger ut5

**Review by mankrit-singh — DISMISSED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by abhinavsingh1196 — APPROVED** [review]
Please consider automation tests for this API in the test plan.

**Review by TarunUba1 — APPROVED** [review]

**Review by tchaurasiya — APPROVED** [review]

**Review by mankrit-singh — APPROVED** [review]

**Review by abosii-harness — APPROVED** [review]

**Comment by GoelPratyush99** [comment]
trigger review
